### ISSUE TITLE
Diagnostics aren't as good as last year for current base model (incl. age-1).

### ISSUE THREAD

**Issue by andrew-edwards** [pacific-hake/hake-assessment#878]
Current base model Appendix A gives this:
![image](https://user-images.githubusercontent.com/6977190/149831765-9b985d12-4996-40d7-b25e-cf99a92269aa.png)

Aaron said the diagnostics looked good when he manually looked at them. Code to make that figures is
```
plot_mcmc_param_stats(base.model)
```
@aaron - did you add other options?

I just had a thought that it may be something to do with the age-1 index now being included, so will take a look. Could be why adnuts-diagnostics document isn't building now.

**Comment by andrew-edwards** [comment]
That Geweke behaviour occurs for the first bridging model ("Update all fishery catch and comps") (but not when I check last year's base model using `plot_mcmc_param_stats()`, ruling out some issue with my setup; I updated r4ss okay though). 
Is that to do with the comments related to e9aa994 ? (Though wouldn't explain the behaviour persisting through to new base model though).

Or now that we have the extra forecast year, maybe some setting somewhere (!) also needs to be changed to account for that.

**Comment by aaronmberger-nwfsc** [comment]
I used r4ss commands:
my_mcmc <-SSgetMCMC(dir ="mymodel", verbose = TRUE, writecsv = TRUE, csv1 = "keyposteriors.csv", csv2 = "nuisanceposteriors.csv", keystrings = c("SSB_2019","Bratio_2019", "SR_LN(R0)", "SR_BH_steep", "RecrDev_2014","EffN_mult"), nuisancestrings = c("Objective_function", "SSB_", "Bratio_", "InitAge","RecrDev"))
pdf("...\\SummaryDiagnostics.pdf")
mcmc.nuisance(dir="...\\models\\",run="mymodel\\",printstats=TRUE)
dev.off()

plot_mcmc_param_stats says it is a recoded version of the mcmc.nuisance function used above.

and got using the base model output on a run I did on my remote desktop (7cores):
![image](https://user-images.githubusercontent.com/7946234/149833855-4e8b97a2-d75b-4ba5-bf56-3aab50b6174d.png)


And the effN seems to match (visually) with that from the Shiny app within adnuts code:


![image](https://user-images.githubusercontent.com/7946234/149832711-ed7a4f36-93ab-4d46-a6c8-874b2d74703e.png)

One thing is to look at is if you run that r4ss code above and look at the table printed to the screen you can see the actual values that go into those plots for each parameter.

**Comment by andrew-edwards** [comment]
Thanks. I thought it might have been a different run of the base model, which makes it curious. Will do what you said after lunch.

**Comment by andrew-edwards** [comment]
@aaronmberger-nwfsc Could you maybe just try `plot_mcmc_param_stats()` with the model you ran (I don't have all the .sso etc. files downloaded, so can't run your code)? Then we'd know if the difference is between the model runs (your computer v AWS) or the code.

**Comment by cgrandin** [comment]
@andrew-edwards `plot_mcmc_param_stats()` takes a model as an argument, you don't need the output files, just load the RDS loaded into ta variable like is done in model-setup.R

**Comment by andrew-edwards** [comment]
Yes, that works fine. I'm talking about Aaron's code above, which directly calls the .sso files. Expect that the difference is with the model runs (not the different analyses in the code), but this will rule out the different analyses fairly easily.

**Comment by cgrandin** [comment]
I think I’m just saying that code is the same as what I’ve done, it’s just in the hake assessment repo it’s done in several steps, i.e create RDS files from output, then run the stats and create the plot. But if you want to run the code from scratch you can download the sso files and other files using the s3_download() function.

**Comment by andrew-edwards** [comment]
Agreed, I just don't think that's the issue and so not worth trying (yet). Here's the bridging figures for the first bridging run:
So the spurious behaviour is happening straight away. Presumably we should try and figure this out first (may also end up helping the retro problem).

![image](https://user-images.githubusercontent.com/6977190/149847984-a1e89d4e-4402-4389-a106-e53dd74f74a0.png)

**Comment by andrew-edwards** [comment]
[Yes, I tweaked my code to make a bridge-diagnostics.pdf for the diagnostics of bridging models. Just don't tell Kelli ;)   ]

**Comment by andrew-edwards** [comment]
Summary:
1. Seems like we're getting different diagnostics for the model run on AWS and the model run by Aaron. 
2. Aaron did the bridging models on his computer and (I'm pretty sure) checked the diagnostics, and they looked fine all the way through to the new base model. 
3. But diagnostics on bridge models and new base (and all sens models) don't look so good, based on the AWS runs.

I've done code to make full diagnostics pdf files for the bridging models and then for the retro models (and the original code does all the sens models). All automated, so we can quickly check them, which should hopefully avoid the current issue.

**Comment by andrew-edwards** [comment]
I started adapting Aaron's code in the above commit, but don't think I downloaded the right .sso files (which is what the .rds stuff is trying to avoid). @aaronmberger-nwfsc tomorrow if you can try `plot_mcmc_param_stats()`  on your locally-run base model, then that will tell us if it is number 1 or 4 in the above list.

**Comment by aaronmberger-nwfsc** [comment]
Yes, I'll run plot_mcmc_param_stats() today.  I did download the hake.RData file and ran the launch_shinyadmb(nuts.updated) from the base model on the drive and manually did a histogram of the effN (below), which seems to indicate there is an issue in our plotting code.  The above task will help further diagnose.
![image](https://user-images.githubusercontent.com/7946234/149970145-f6e5f87e-f7ac-42c9-9a58-57c584ae80dc.png)

**Comment by andrew-edwards** [comment]
Here's the plot for Chris's `2021.00.04_base_2022_code`, which is last year's base model using this year's setup. So looks different to plotting last year's base model from last year's runs, implying it's number 1 in my list above (to do with AWS run, or the automation) rather than 4 (the plotting function).

![image](https://user-images.githubusercontent.com/6977190/150020864-057b76aa-9af8-437b-ad96-ee96d0c6c7a2.png)

**Comment by andrew-edwards** [comment]
5. We could run the base model using the automated functions on a local machine (like last year), to tease out whether it's AWS v local machine, or the functions.

**Comment by cgrandin** [comment]
If you want to compare machine output, you need to compile optimized SS at commit a0f982d9258e626c63d69ba3366e99b9744d0cc4

https://github.com/nmfs-stock-synthesis/stock-synthesis/tree/a0f982d9258e626c63d69ba3366e99b9744d0cc4

**Comment by aaronmberger-nwfsc** [comment]
And remember we still have a difference between the effN in the upper right of these plots and the shinyapp for the same exact model run on aws.

**Comment by cgrandin** [comment]
@aaronmberger-nwfsc that's what makes me think it's just plotting code

**Comment by andrew-edwards** [comment]
But if it's just plotting code, why would it work okay for last year's base model (as run last year):
```
plot_mcmc_param_stats(last.yr.base.model)
```
![image](https://user-images.githubusercontent.com/6977190/150023263-fe1b5ac8-1cae-4a48-a530-f49f6229aa48.png)

which matches last year's Figure A.4. And doesn't match the re-run of last year's model (`2021.00.04_base_2022_code` shown earlier).

**Comment by andrew-edwards** [comment]
@aaronmberger-nwfsc - what was your `adapt_delta` on your locally-run model? 
Chris used 0.95 for all (yes, Chris?).

**Comment by cgrandin** [comment]
Yes, 0.95 for all models including the re-run of last year's base model.

**Comment by andrew-edwards** [comment]
Aha - but last year's had 0.90 yes?

**Comment by cgrandin** [comment]
Yes it did

**Comment by cgrandin** [comment]
Here is last year's using 0.95 and this year's model code and everything:

![image](https://user-images.githubusercontent.com/4551051/150025375-8c5d4c0d-fed8-4a21-9011-ae0fc7a94704.png)

**Comment by andrew-edwards** [comment]
Yep, same as my earlier plot (36 mins ago).

**Comment by cgrandin** [comment]
Here is this year's base run at 0.95 with 12 CPUs only (instead of 16 which is what all the rest are)

![image](https://user-images.githubusercontent.com/4551051/150025621-bc4e57b3-0ae8-4fdf-804c-fac8541f03a4.png)

**Comment by andrew-edwards** [comment]
Do you want to run that but with adapt_delta = 0.9?

**Comment by cgrandin** [comment]
And finally, this year's with 16 CPUs and 0.95 (what we want in the document)

![image](https://user-images.githubusercontent.com/4551051/150025805-87fb2d6e-cf5c-4573-b176-b95b22f9f502.png)

**Comment by andrew-edwards** [comment]
Yep, this matches the figure at the top of this Issue. So think we need to try `adapt_delta = 0.90` again.

**Comment by cgrandin** [comment]
OK I can run that now

**Comment by aaronmberger-nwfsc** [comment]
I'm still checking, but I think this has something to do with altering the type of output we want in our report files.  For example, we now output 124 more derived outputs relative to dynamic B0 and dynamic recruitment.  As well as few refinements with selectivity - although that just got rid of some parameters that were fixed so shouldn't have been reporting them.  If the dynamic B0 stuff is included then I can see that being the issue we are seeing.  I'm going to dive into the details more in plot_mcmc_param_stats() to find out.  At the end of the day, we really only need to be reporting the performance of active parameters in these diagnostics.  I don't see a reason to report derived quantities, since those are a function of the active parameters.

@kellijohnson-NOAA  thoughts?

**Comment by andrew-edwards** [comment]
Thanks. And from Slack on 13th Jan Chris said: "I started running all the models in parallel last night (late) so they are still going. I forgot to set adapt_delta = 0.9 for the bridge models and sensitivities so they are running with adapt_delta = 0.95."  Which explains why the very first bridging model shows the spurious behaviour.

**Comment by aaronmberger-nwfsc** [comment]
So it may not be an issue per se, just that we are plotting things we don't need to... still looking...

**Comment by andrew-edwards** [comment]
@aaronmberger-nwfsc  - yes, that's another change since last year. Chris's run will see if it's adapt_delta or not (if it is we'd have to re-run stuff, boo.). 
And your investigations will figure out if we're just including too many things in the plotting - the actual plotting functions seem to work fine....
We're also doing the extra forecasting year, but I don't think that would necessarily be it.

**Comment by cgrandin** [comment]
@aaronmberger-nwfsc 
I looked more closely at the code you wrote above to generate the initial plot that's different than what we have in the document:

```
my_mcmc <-SSgetMCMC(dir ="models/2022.01.10_base/mcmc/sso", verbose = TRUE, writecsv = TRUE, csv1 = "keyposteriors.csv", csv2 = "nuisanceposteriors.csv", keystrings = c("SSB_2019","Bratio_2019", "SR_LN(R0)", "SR_BH_steep", "RecrDev_2014","EffN_mult"), nuisancestrings = c("Objective_function", "SSB_", "Bratio_", "InitAge","RecrDev"))
pdf("...\SummaryDiagnostics.pdf")
mcmc.nuisance(dir="...\models\",run="mymodel\",printstats=TRUE)
dev.off()

```

You create an object called `my_mcmc` which is then never used anywhere. So the `mcmc.nuisance()` function is the only thing run here to produce the plot.

I ran the same thing on the current base model:
`mcmc.nuisance("models", "2022.01.10_base/mcmc/sso", printstats = TRUE)`
and get this:

mcmc.nuisance("models", "2022.01.10_base/mcmc/sso", printstats = TRUE)

So they match perfectly.

**Comment by cgrandin** [comment]
So you need to run `build_rds("your-model-run-dir")` and then 

```
model <- load_models("your-model-run-dir")
plot_mcmc_param_stats(model)
```

**Comment by andrew-edwards** [comment]
@cgrandin  - no point in that I think, as I think we've ruled out differences in plotting code. Your previous post just confirmed it.

**Comment by andrew-edwards** [comment]
@aaronmberger-nwfsc 
`plot_mcmc_param_stats` uses `parm_nm <- model$post_names`. 
Here are the differences in `post_names` between last year's base model and this year's:
```
> length(base.model$post_names)
> [1] 251
> length(last.yr.base.model$post_names)
> [1] 243
> setdiff(base.model$post_names, last.yr.base.model$post_names)
 [1] "NatM_uniform_Fem_GP_1"            "Main_RecrDev_2019"               
 [3] "Main_RecrDev_2020"                "Late_RecrDev_2021"               
 [5] "ForeRecr_2024"                    "ForeRecr_2025"                   
 [7] "Q_extraSD_Age1_Survey(3)"         "AgeSel_P3_Fishery(1)_DEVadd_2021"
 [9] "AgeSel_P4_Fishery(1)_DEVadd_2021" "AgeSel_P5_Fishery(1)_DEVadd_2021"
[11] "AgeSel_P6_Fishery(1)_DEVadd_2021" "AgeSel_P7_Fishery(1)_DEVadd_2021"
```
You mentioned 124 extra derived outputs, but they don't seem to be included here.

**Comment by andrew-edwards** [comment]
Ooh, but 10 of those extra ones do get added on in Chris's re-run of last year's base model:
```
> setdiff(base.model$post_names, last.yr.base.this.yr.code$post_names)
 [1] "Main_RecrDev_2019"                "Main_RecrDev_2020"               
 [3] "Late_RecrDev_2021"                "ForeRecr_2025"                   
 [5] "Q_extraSD_Age1_Survey(3)"         "AgeSel_P3_Fishery(1)_DEVadd_2021"
 [7] "AgeSel_P4_Fishery(1)_DEVadd_2021" "AgeSel_P5_Fishery(1)_DEVadd_2021"
 [9] "AgeSel_P6_Fishery(1)_DEVadd_2021" "AgeSel_P7_Fishery(1)_DEVadd_2021"
```

**Comment by cgrandin** [comment]
I think it's just because we used 0.9 last year and 0.95 this year. I'm running this year's base at 0.9 to find out for sure.

**Comment by cgrandin** [comment]
@andrew-edwards some of those are the same as last year, just incremented year by one. i.e replace the 2021's with 2020 and they will be in there last year

**Comment by aaronmberger-nwfsc** [comment]
0.95 this year will result in less divergences than the same model used with 0.90 so the effective N would increase, but we are seeing decreases in effective N, so I don't think it's the adapt_delta.  Maybe I'll be proven wrong...

**Comment by cgrandin** [comment]
We should know after dinner today :)

**Comment by andrew-edwards** [comment]
Think I should have done: 
```
> setdiff(last.yr.base.model$post_names,` last.yr.base.this.yr.code$post_names)  # in last year's base, but not in this year's re-run of the base
[1] "NatM_p_1_Fem_GP_1"
> setdiff(last.yr.base.this.yr.code$post_names, last.yr.base.model$post_names)   # vice versa
[1] "NatM_uniform_Fem_GP_1" "ForeRecr_2024"        
```
But taking those out in a hack of `plot_mcmc_param_stats()` doesn't fix it.

**Comment by cgrandin** [comment]
Oh, I didn't realize `NatM_p_1_Fem_GP_1` changed to `NatM_uniform_Fem_GP_1`. Thanks. That will probably cause some issues with natural mortality plots

**Comment by cgrandin** [comment]
This is this year's model run with `adapt_delta = 0.9`

![image](https://user-images.githubusercontent.com/4551051/150062833-9523df35-99a2-4a6f-b638-9ef046fc2313.png)

**Comment by cgrandin** [comment]
So the only thing left is that something changed in the SS code

**Comment by aaronmberger-nwfsc** [comment]
Interesting:  I used our aws base model posteriors.sso and derived_posteriors.sso and compared those to the same files from the base model I ran locally (these are the two versions that give very different diagnostic results) and I got nearly identical median results (identical for all practical purposes given they were different number of chains).
![image](https://user-images.githubusercontent.com/7946234/150065606-649d2a64-123a-4a62-ae8f-c376976edac3.png)
![image](https://user-images.githubusercontent.com/7946234/150065666-62b3dc85-4f52-4e1c-be45-b5910ad94382.png)

Why would 'identical' results give such different diagnostics???

**Comment by cgrandin** [comment]
I think maybe we should start by comparing last year's model rds file and last year's model run with this year's code since they are different and should be identical.

 I'm debugging the plotting function with those in an attempt to understand

**Comment by cgrandin** [comment]
But that is interenting, and doesn't make immediate sense!

**Comment by cgrandin** [comment]
This is the effN's from lasy year's model run with this year's code:
```
  [1] 3699 1348 8000 6552 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 7354 7128 7777 7272 7385 6464 6149 3593
 [26] 4857 6778 1907 2819 3315 1697 3959 1822 3678 1421 4100 3269 1199 4519 3183 4088 1106 4519 3963 1193 1783 4186 1152 2538 4520
 [51] 1144 1131 1532 1304 1717 1266  976 3507 1112 4616 1032 3889 1068 1131 4574 1085 1906 1200 3178 1608 3564 1622 4176 2483 3887
 [76] 6875 8000 8000 8000 8000 8000 8000 7096 6484 6441 7373 4364 3530 8000 8000 7955 8000 8000 7678 8000 8000 8000 8000 8000 8000
[101] 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 8000 7444 8000 8000 8000 8000 8000 8000 6949 8000 8000 8000
[126] 8000 8000 8000 8000 8000 8000 8000 7873 8000 8000 8000 8000 8000 8000 7953 8000 8000 8000 8000 7267 7431 8000 7078 8000 7184
[151] 8000 8000 7696 8000 8000 8000 8000 8000 8000 8000 8000 8000 7182 7901 6970 8000 8000 6825 8000 8000 8000 8000 8000 8000 8000
[176] 7523 6557 6611 7231 7555 6993 8000 8000 6562 8000 8000 8000 8000 8000 8000 8000 8000 7288 7526 6427 7492 7340 6134 7551 8000
[201] 8000 8000 8000 6900 8000 8000 6448 6770 6764 6481 7116 7896 8000 8000 8000 8000 7696 8000 8000 8000 8000 8000 8000 6858 7975
[226] 7694 7772 7410 8000 7979 8000 8000 8000 8000 8000 8000 7578 7036 7186 6862 7817 7888 6303 8000
```

And last year's model with last year's code:
```
  [1] 4571 3534 2124 8250 8250 8250 7555 8250 5939 5022 8250 7258 8034 8250 8250 7040 5349 5263 6796 6218 5442 3339 3761 2356 2280
 [26] 2350 3766 4632 3582 2586 4080 3786 4568 3000 3636 4514 5351 3589 3297 3636 4043 3403 4653 3930 3614 4170 3614 3621 3849 3823
 [51] 3568 3957 4492 4353 5345 4815 3893 4074 4300 4495 3897 2279 3965 4341 4374 3989 5499 4292 5175 4650 3695 4322 4162 4869 6535
 [76] 5101 8250 8250 8127 8250 7598 5857 7158 8250 7973 8250 8250 8250 7421 8250 8250 7166 5353 7793 4582 8250 8250 7314 7173 5117
[101] 8250 7364 6470 8250 8250 8250 6007 8250 5234 7753 8250 7594 7498 6742 7669 8250 8250 6757 8149 8250 7510 7375 8250 8250 8250
[126] 6620 6947 7178 6760 6425 6316 7453 7508 6526 8250 7040 7458 8250 5646 7438 8250 8050 7148 6493 8250 6412 6963 6782 5210 8250
[151] 7896 8250 6828 6398 8250 8250 6987 8250 6742 8250 8250 8250 6888 8250 8250 5734 4333 8046 8250 7699 7245 8250 7777 8250 8250
[176] 7601 5149 5253 5265 6973 8046 5969 6582 8250 7112 8250 8250 6702 8250 8250 8250 8250 6883 8250 8250 6965 4001 5820 8250 7760
[201] 6323 8250 6857 8250 8250 7751 6526 4959 7334 6881 8039 5753 7033 8250 7172 6500 8250 8250 6784 8250 8250 8240 6110 8250 7540
[226] 8250 8250 4526 8250 7927 8250 8250 6440 8250 8250 8250 7504 8250 7506 8250 8050 8250 6717
```

Almost every 8000 this year was 8250 last year

**Comment by cgrandin** [comment]
This is stats$autocor > 0.4 for the new code run:

```
  [1] FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [21] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE
 [41] FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE  TRUE  TRUE FALSE  TRUE FALSE  TRUE  TRUE FALSE  TRUE FALSE
 [61]  TRUE FALSE  TRUE  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [81] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[101] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[121] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[141] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[161] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[181] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[201] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[221] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[241] FALSE FALSE FALSE FALSE
```

This is why there are "No test" on the new plots. There is autocorrelation early on

**Comment by aaronmberger-nwfsc** [comment]
The 8250 versus 8000 is just because last year 8250 was the max because I ran the base model with a slightly different iteration setting.  

I ran the base model on my laptop last night using 11 cores and got the same 'better' diagnostics to my runs using 7 cores. The only differences from the models run through aws by Chris are 1) I used the ss.exe from late December that Kelli gave me, 2) I used the function adnuts-run-base-model.R (rather than run-adnuts.R), and 3) I ran it on a windows machine rather than a linux machine.  

None of these differences make clear sense to me as to why the diagnostics are different (and further, that the median estimates for derived quantities and the median parameter estimates themselves are nearly identical between the two versions with such different diagnostics).  The effective sample size plot I get when running launch_shinyadmb(nuts.updated) on the hake.Rdata file produced in the adnuts code matches what I get with the 'better' diagnostic plots. This is the case for when I load hake.Rdata using s3_download(file.path("models/2022.01.10_base/mcmc", c("hake.Rdata")))
from the aws run and when I load it from my local machine run (see below).  But yet our 4-panel diagnostic plots are different. 

![image](https://user-images.githubusercontent.com/7946234/150177168-4c088fa1-7872-4829-adc6-efbe00cba566.png)

Complete overlap.  So I don't understand why the diagnostics (e.g., effective sample size) are different in those plots.

**Comment by andrew-edwards** [comment]
Good work. So it sounds like an issue in the plotting code, given the detective work you've both done, yes?

@aaronmberger-nwfsc - did you make an .rds from one of your local model runs, and run  `plot_mcmc_param_stats(model)`?  (Looks like I said not to bother yesterday as I thought it was `adapt_delta`, but Chris has ruled that out.

The other thing is it could be to do with how the .rds file is being built. 

I think the 'diagnostics being different but biomasses etc. being similar' issue could just be to do with including extra un-needed parameters in the diagnostics for the four-panel plot (or in the .rds file). So the 'true diagnostics' are fine (as Aaron has found manually), but we're just showing too much.

@cgrandin - note that in your "This is the effN's from lasy year's model run with this year's code:" post above, the second set of effN's (from last year's model with last year's code) has one extra value. Not that that should be enough to mess everything up though, but it means something is slightly different.

**Comment by cgrandin** [comment]
@andrew-edwards I rewound everything back and created a new Docker container which holds all the code in last year's state, and ran last year's model on AWS. It looks as it should:

![image](https://user-images.githubusercontent.com/4551051/150180235-35fddf70-2289-4f18-be93-e17b0ca6f77d.png)

So the final thing to try is running this year's code but using last year's SS. I am creating a Docker image to do that now. My guess is we will find something changed in SS over the year. What is puzzling is that we cannot figure out why from the outputs.

**Comment by andrew-edwards** [comment]
Aarons was using SS from 24 Dec, Chris was building the latest version. So Kelli and Chris investigating further as it seems to be due to changes between those commits .....

**Comment by andrew-edwards** [comment]
I'm not worrying/thinking about actual results yet, but a heads up that median B0 has come out as 900,000 t compared to last year's 1,658,000 t (and last year's age-1 was around that latter value also). 

![image](https://user-images.githubusercontent.com/6977190/150209015-8ac692b2-ca1c-4534-9bd8-21eb21c42e11.png)

And:
```
> curr.bio.median.tonnes
     2022 
"618,535" 
```
But this is not what Aaron found in his above plot of comparisons of SSB. Is this the divide by 2 issue (I thought that was only for total biomass)?

**Comment by aaronmberger-nwfsc** [comment]
In my MLE runs, adding the age 1 index increased B0 some compared to last year (see below), which makes me think there is a lingering SSB/2 lingering in there.
![compare1_spawnbio](https://user-images.githubusercontent.com/7946234/150209768-8413ab14-d8cf-4ab4-97a4-10ae18f9128f.png)

**Comment by cgrandin** [comment]
We still haven't divided last year's by 2. In the RDS file

**Comment by andrew-edwards** [comment]
No, last year's is correct. Currently we have:
```
> last.yr.base.model$mcmccalcs$smed[c(55:57, 59)]
     2020      2021      2022  unfished 
1.2995225 0.9808500 0.7169125 1.6577300 
> base.model$mcmccalcs$smed[c(55:57, 61)]
    2020     2021     2022 unfished 
0.764510 0.667555 0.618535 0.903665 
```
(I posted in this Issue as wondering if it was sampling some low Biomass values that may be related to this Issue, but it's probably /2).

**Comment by cgrandin** [comment]
There's nothing touching that value in our code. Look at last year's table:

![image](https://user-images.githubusercontent.com/4551051/150211602-cb1e5337-0bd5-41a9-9433-43d0cbb57696.png)

**Comment by cgrandin** [comment]
OK, I did a check on the actual posterior and it is 1,807

**Comment by andrew-edwards** [comment]
The current equivalent to Aaron's above plot is:
![image](https://user-images.githubusercontent.com/6977190/150212247-fa9dbd43-fd35-448c-b6a1-d0f5050c11d7.png)
So the 2021 base model is **now** showing twice the biomass that it did last year. This should be a new Issue then.

**Comment by cgrandin** [comment]
I fixed it in the table

**Comment by cgrandin** [comment]
This discussion about halved biomass should be in another issue

**Comment by andrew-edwards** [comment]
So, regarding diagnostics:
1. On AWS Chris ran last year's model with last year's SS, diagnostics look good (like we had in last year's assessment).
2. On AWS Chris ran this year's model with 24 Dec 2021 SS, but diagnostics bad.
3. Aaron ran the current base model on his laptop last night using 11 cores and 7 cores and gets good diagnostics. Using the SS from late December that Kelli gave him (presumably 24 Dec).

**Comment by cgrandin** [comment]
I'm wondering - @aaronmberger-nwfsc are you 100% sure you are using the newer SS (dec 24, 2021)? I'm wondering of there is a chance it is last year's executable

**Comment by aaronmberger-nwfsc** [comment]
Last years ss.exe was 7,023 KB and the one I'm using is 7,126 KB and it is the version that Kelli compiled for me and sent to me via google drive on December 23 (I downloaded it on December 24 and thus that time stamp).  I don't know if the compile date was exactly December 23 (or 22) but it would have presumably been right in that time frame.

**Comment by cgrandin** [comment]
Oh! So you have one compiled BEFORE Dec 24. I compiled at the commit that happened on Dec 24 and tested that. We need to know exactly which commit the one you are using was compiled at.

**Comment by kellijohnson-NOAA** [comment]
So, after many iterations and chatting with Rick he does not believe that this is something that we should be spending our time on. 
```
[In MLE] this is just a slight difference in N iterations before convergence; hence tiny difference in final result for the MLE run.  Amazing that it blows up into something noticeable in the chain.  Butterflies flutter in China and change Seattle weather.......
```

**Comment by cgrandin** [comment]
@kellijohnson-NOAA does than mean, accept the diagnostics as they are and move on?

**Comment by aaronmberger-nwfsc** [comment]
Sorry, I don't follow what Rick is saying here.

**Comment by andrew-edwards** [comment]
But don't we need to figure out if it's
a) the diagnostic plots and analyses are wrong (somehow including extra variables that shouldn't be included, maybe due to SS output), or
b) the diagnostic plots are right and so there'd be issues with the results?

**Comment by kellijohnson-NOAA** [comment]
For completeness ... here is the "linear" commit history of SS3 where the hard return is prior to the commit that led to changes in the order of operations and thus differences in convergence.
```
a0f982d fix issue with re-opening of report.sso iostream
48fd7d7 Merge branch 'main' of https://github.com/nmfs-stock-synthesis/stock-synthesis into main
920f7fc remove unneeded debugging warning in popdyn
52387e8 Merge pull request #250 from nmfs-stock-synthesis/improve_procedure_order_of_operations
86ee785 Merge pull request #241 from nmfs-stock-synthesis/spline_issue_86
56e6c78 Merge pull request #249 from nmfs-stock-synthesis/Z_by_area_r4ss_changes
6bc19ee Merge pull request #218 from nmfs-stock-synthesis/output_to_separate_files
98deac0 Merge pull request #248 from nmfs-stock-synthesis/fix_WTATAGE_WIP
69dbf86 Merge pull request #244 from nmfs-stock-synthesis/augment_Fmethod_4_#243
698a53f Merge pull request #237 from nmfs-stock-synthesis/gha
d8c2fdd Merge pull request #236 from nmfs-stock-synthesis/fix_issue#205_for_seasonal_forecast

1caa88f Merge pull request #233 from nmfs-stock-synthesis/improve_predator_M2
60492ac testing, #231: rm jenkins workflows
f07a63d Merge pull request #230 from nmfs-stock-synthesis/replace_IF_with_lookup
a8d3f9f Merge pull request #223 from nmfs-stock-synthesis/change_abbrev
aa16683 updating
a7a1321 Merge pull request #225 from nmfs-stock-synthesis/Andre_bioecon
b697618 Merge pull request #229 from nmfs-stock-synthesis/r4ss_tests
a600337 one more fix
```

ADMB message related to convergence
```
ic > imax  in fminim is answer attained ?
Function minimizer not making progress ... is minimum attained?  
```

Version 3.30.18
```
 Iterations: 727 -log(L): 2028.65
Final gradient: 0.000167504
```
Version 3.30.19
```
 Iterations: 730 -log(L): 2028.65
Final gradient: 0.000106248
```

**Comment by aaronmberger-nwfsc** [comment]
Okay, very interesting.  I just ran our base model on leafcutter (Kelli and my server cpu) using the ss_a0f982d.exe with the following adnuts setting: 
adapt_delta = 0.95
iterations = 12,000
cores = 12
warmup = 500

And here is what I got (the parameters shown in the first plot are steepness, ln(R0), and M: 

![image](https://user-images.githubusercontent.com/7946234/150623268-cf39c830-78d2-419a-98df-56af072dcfa2.png)

![image](https://user-images.githubusercontent.com/7946234/150623283-731096db-7fa9-4c61-9259-3e605c8e2f38.png)

![image](https://user-images.githubusercontent.com/7946234/150623347-4b1dda94-9e8a-4719-bf5a-8054ec3b0cf2.png)

Look at the threshold % for the effective sample size - no parameters less than 25% (3 if you go to 30%)

What am I missing?  I think we may have been having a burn in (warmup) issue with the non-zero constrained rec devs, but after sufficient warmup they settle in better.  I have another run starting that is on 24 cores with a warmup of 500 and 16,000 iterations that will run overnight.  Thoughts?

**Comment by aaronmberger-nwfsc** [comment]
Same thing with a 24 core machine and 16,000 iterations with 500 warmup. Except with more cores (i.e., smaller length of chain/core): there was one rec dev parameter below a 25% effN threshold.

![image](https://user-images.githubusercontent.com/7946234/150646514-a73f883a-ed54-403e-9dd9-86a2406c7c89.png)

![image](https://user-images.githubusercontent.com/7946234/150646590-28c59fc0-d2a0-4f21-95ef-9598d3321158.png)

![image](https://user-images.githubusercontent.com/7946234/150646625-5d10740d-fa5e-421c-9ce8-476ae6ad4826.png)

**Comment by kellijohnson-NOAA** [comment]
Are you thinking that the inference of the sensitivities are still okay but that we should use longer chains with a longer warmup period for the base model?

**Comment by aaronmberger-nwfsc** [comment]
I guess there are 2 things I think we should do:

- [ ] check if these runs only really help with autocorrelation (i.e., increase effN) but really make no meaningful difference in inference (for reasons we discussed yesterday) - if so then yes, I think all the other sensitivity runs are okay. @aaronmberger-nwfsc  to do
- [ ] I'd like Chris to send off an exact same specification run using his workflow, because given mix ups before I just want someone else to duplicate this general finding (less autocorrelation with new run specifications).  So this would be running one with 500 warmup (warmup_final), 12,000 iterations (n_final) on 12 cores (reps) using adapt_delta=0.95.  The run itself took ~3 hours. @cgrandin to do

Pending the above we can decide if we want to update the base model rds or not. I still think our ideas for changes in diagnostic plots are still valid changes regardless (i.e., show rhat and effN as summary measures; and the parameter specific quad plots for key parameter plots).

**Comment by aaronmberger-nwfsc** [comment]
Regarding check box number one above - looks like we are not worried about inference, it is just in the effN.  Shown below is our base model that we have decided to use (labeled aws here) and the model I ran and referenced above (with 12,000 iterations, etc.; labeled leaf here). The leaf run (blue line) does have the smallest expansion in the 95% confidence interval (dashed lines) as would be expected with considerable more effective samples, and that small expansion increases nearer the end of the time series, but again the main inference is almost identical.
![SSB_comprare](https://user-images.githubusercontent.com/7946234/150652567-36d25570-e06b-491a-b613-11337e1bb7e1.png)

And parameter median comparison:

![Median](https://user-images.githubusercontent.com/7946234/150652599-9a4093c8-ee52-4ed7-85c8-934bf0a807e2.png)

And parameter SD:

![SD](https://user-images.githubusercontent.com/7946234/150652612-dc7516ee-c4ee-4022-a269-5882abc4b63d.png)

Bottom line is that we are probably just fine using the base model we have now (with not ideal but okay diagnostics), but if we wanted to be extra 'tight' and having everything 'look great' upon first glance we could rerun the base model with these updated adnuts input parameters.

**Comment by aaronmberger-nwfsc** [comment]
Regarding check box 2 above:

@cgrandin : can you send off a run to AWS please.  I want to double check that what I've done comes out the same using your workflow.  I'd like the run to be the following:
-Model: Base model
-adapt_delta: 0.95
-iterations: 12000
-warmup: 500
-cores: 12
-ss.exe: ss_a0f982d.exe (the one you have been using)
Everything else can be the same.
I ran another set and got the same results b/c I wanted to check that using adnuts-run-base-model.R resulted in anything different that using run-adnuts.R - it didn't really, but I had to run the latter line by line because it would error out for me relative to the hess_step stuff.  While doing so I noticed the following things (which I don't think any should matter)
-sample_admb should replaced throughout (2 spots) with sample_nuts per Cole's guidance (sample_admb is deprecated)
-when I ran the code it conducted the mceval stuff twice (once as part of the nuts_updated call and then again in the subsequent system call) - though maybe that is necessary for the mcmc_path stuff.
I'll copy this to a new comment in issue#878 also. (edited)

**Comment by aaronmberger-nwfsc** [comment]
and also build the rds file for the above model

**Comment by cgrandin** [comment]
I have started the run...  I hadn't read this today as I wasn't feeling all that great.

It's interesting though that more samples with  a little larger warmup would improve things so much.
More to come..

**Comment by cgrandin** [comment]
I also set off another one for interest's sake, identical to your run with 24 cores but with 36 cores:
CPUS: 36
adapt_delta: 0.95
n_final: 16,000
warmup_final: 500

**Comment by cgrandin** [comment]
Note that when I say CPUS, they are actual CPUS, not hyperthreaded. For example my laptop has 4 CPUs but they advertise it as having 8 "cores" and when you look at task manager, there are 8 cores. There are only actually 4 CPUs but they are "hyperthreaded". I'm just checking with you that you are using actual CPUs in your core numbers and not including hyperthreaded cores.

**Comment by aaronmberger-nwfsc** [comment]
I'm reporting numbers that pertain to detectCores(), and really I'm only interested in the number of chains produced, whether they are actual CPUs or hyperthreads doesn't matter for my purposes anyway.  I know it likely matters for speed.

**Comment by cgrandin** [comment]
That's true, the number of cores = the number of chains is for the  `run_adnuts()` function so I guess it doesn't really matter. I was thinking there may be halving of number of chains or something but no there isn't.

**Comment by kellijohnson-NOAA** [comment]
@cgrandin I think we can close this one right?

**Comment by cgrandin** [comment]
Yes! Done with this one
